### PR TITLE
DM-2176: Add explore all practices page

### PR DESCRIPTION
**Pull request by emyl3**
### JIRA issue link
[DM-2176](https://agile6.atlassian.net/browse/DM-2176)

## Description - what does this code do?
This PR adds functionality for the "Explore all practices" page.
In addition, this PR also addresses the following:
- practice card styling changes - on desktop view, practice cards take up full width of the container
- scss color variable update
- caches practices based on 3 sorting options
- resets the cache when a `CategoryPractice` and `DiffusionHistory` are updated/deleted

## Testing done - how did you test it/steps on how can another person can test it 
**Explore all practices**
1. Visit `/explore`
2. Ensure you do not see any breadcrumbs from previous site navigation
3. When first loading or reloading the "Explore all practices" page, check the following:
- no categories are selected
- the sort option is set to "Sort by A to Z"
- the total results count = total number of published practices in your local DB
- only first 12 practices in ABC order should be displayed
- if you have more than 12 practices, there is a load more button
- "Other" and subcategories of the "Other" category are not displayed as a filter option
- Categories that have not been selelcted by a practice are not displayed as a filter option
4. After clicking a category, check the following:
- a page spinner appears where the practice cards were (the results count and any load more buttons should not be visible)
- when the results load, the practice should be sorted based on the selected sorting option
- when selecting other categories, the results should be additive and the results count should update accordingly
- for category filter selection with more than 12 results, there should be a load more button
- changing the sort option only sorts the practices that are filtered out by then categories filter
- NOTE: if a practice is a part of two or multiple categories that you selected, it is only counted once towards the total results count
5. After selecting a new sort option, check the following:
- a page spinner appears where the practice cards were (the results count and any load more buttons should not be visible)
- the results count remain the same when only sorting the results
6. After clicking the "Load More" button, check the following:
- the remaining results should also be sorted according to the sort option selected
- the results are loaded 12 at a time
- when the final practices have all been loaded, there should be no "Load more" button
- If you have more than 24 practices, there should be another "Load more" button and so forth...
7. On the mobile version, check the following:
- if you have more than 5 categories, there is a "See all categories" link button
- When clicking the "See all categories" link button, it displays all the remaining categories
BONUS:
- In the code, go to the `app/views/practices/_explore.js.erb` file.
- Go to line `96`
- change the `url`value to something completely wrong like `/explore-bad-path`
- Reload the page
- Select a new sort option
- You should see an error message (see screenshots below)

**Practices caching**
1. Clear the cache in the rails console:`Rails.cache.clear`
2. Check that the keys don't exist: `Rails.cache.redis.keys`
3. Go to `/` or `/explore`
4. Check that the `searchable_practices_a_to_z` key now exists
5. Go to `/explore`
6. Select the other two sorting options
7. Check that the `searchable_practices_added` `searchable_practices_adoptions` keys now exists
8. Edit a practice and save
9. Check that the keys don't exist: `Rails.cache.redis.keys`
10. Repeat steps 3-7
11. Remove or add a category to a practice
12. Repeat steps 3-7
13. Remove, add, or edit a diffusion history for a practice
14. Repeat steps 3-7

**Practice card styling**
1. Go to various pages where there are practice cards (e.g. Profile, Recommend for you, page builder pages with practices, explore all practices, home page, etc...)
2. Ensure the cards are to-design
3. Most noticeable would be home page and explore all practices page (the row of practice cards width matches the other elements on that page that take up the full width of the container)

## Screenshots, Gifs, Videos from application (if applicable)
**Explore all practices (default)**
Screen Shot 2020-12-01 at 5 21 55 PM

**Load more**
Screen Shot 2020-12-01 at 5 23 33 PM

**Error Message**
Screen Shot 2020-12-01 at 5 22 27 PM

## Link to mock-ups/mock ups (image file if you have it) (if applicable)
[Figma](https://www.figma.com/file/y9U9oLXTGe24O9W3X5UoYW/Home-%2B-Search-V2?node-id=480%3A17203)

## Acceptance criteria
- [X] Build mobile first
- [X] User can select a category
- [X] List of categories is here (didn't write a script since this is already on prod)  
- [X] Additive search results when multiple categories are selected
- [X] Initial display loads 12
- [X] Initial display is A-Z
- [X] Load More button loads 12 more
- [X] Sorting updates practices immediately
- [X] If navigate away and come back, none of filter or sorts are still there
- [X] If you go to select a different dropdown, there’s a check next to what is selected and hover has blue color
- [X] Sorting: Most Adoptions, A-Z, Date Added
- [X] Use new handoff flow
- [X] URL is “/explore”

## Definition of done
- [ ] Unit tests written (if applicable)
- [X] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [X] A link has been provided to the originating JIRA issue
- [X] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/views/practices/_explore.js.erb [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    headers: {
+      'X-CSRF-Token': $('meta[name="csrf-token"]').attr('content')
+    }
+  });
+}
+
+function sendAjaxRequest(data) {
+  var countText;
+
+  $.ajax({
```
Comment (emyl3) [inline app/views/practices/_explore.js.erb:92]: ⚠️ Let me know if there is a more "rails"y way to achieve this 😅

File: app/models/practice.rb [hunk 2]
```diff
@@ -50,9 +51,19 @@ def reset_searchable_practices
     clear_searchable_cache if self.reset_searchable_cache
   end
 
-  def self.searchable_practices
-    Rails.cache.fetch('searchable_practices') do
-      Practice.get_with_categories
+  def self.searchable_practices(sort = 'a_to_z')
```
Comment (emyl3) [inline app/models/practice.rb:54]: ⚠️ Not sure if it is better to cache these sorted results leveraging ActiveRecord or just get the result of `Practice.get_with_categories_and_adoptions_ct` and then sort the results with ruby?? Let me know what you all think 🤔

**Review by JoshingYou1 — COMMENTED** [review]
Awesome work on this Elisa! Everything worked as intended programmatically. Only noticed a few small things.

1. When visiting `/explore` as a logged in user, the bookmark icons for the first three practices is moved down a bit.
Screen Shot 2020-12-03 at 10 52 10 AM
2. Not sure if this has to do with my DB or not, but the count for published practices on the `explore` page was one less than my DB count.
Screen Shot 2020-12-03 at 10 39 50 AM
Screen Shot 2020-12-03 at 10 40 05 AM

3. When visiting `explore` page with a mobile viewport, `320px` in particular, the practice cards seem to stretch past the field of view.
Screen Shot 2020-12-03 at 11 14 33 AM

Once again, great job on this!! :grin:

**Comment by bradjohnson92008** [comment]
FYI - Practices that display in the explore all pages section must have all three of these set to true: 'enabled', 'approved' and 'published' ...  I had old practices that enabled and published but not approved - bc they did not have a user email associated with it.  but all practices where all three vars are set to true are displaying in the explore section..

select count(*) from practices where enabled=true AND published=true AND approved=true

**Comment by bradjohnson92008** [comment]
Hey Elisa - Great job - this was a lot of work.  Regarding "Categories that have not been selected by a practice are not displayed as a filter option"
The number of categories being displayed for me does not match those selected by the practices displayed.  I have a practice that is not published - so it is not displaying in the list of practices.. but it is the only practice I have associated with "Womes Health" ... and the Womens health category is being presented as a filter. 
image

**Comment by bradjohnson92008** [comment]
After clicking a category or changing the sort option I still see the previous results count.
image

**Review by bradjohnson92008 — CHANGES_REQUESTED** [review]
Great work Elisa.  See previous comments I left.  Only other recommendation I have.. is would it be possible to disable the category buttons.. once you click on one to filter?  In my ENV I see the spinner for 5 to 10 seconds.. and I did notice that I could continue clicking on other categories while the app is already working on the first request...  not a big deal if not.

**Review by bradjohnson92008 — APPROVED** [review]
Both issues I found have been addressed - result count not disappearing when clicking a category to filter as well as the Categories displayed are exactly the categories that are associated with my approved/enabled/published practices.  :)  Great work Elisa!

**Review by JoshingYou1 — APPROVED** [review]
Looks and works great. Nice job Elisa!!

**Pull request merged**